hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    };
 
-int halError(uint32_t error) {
-    switch (error) {
-    case NRF_ERROR_INVALID_STATE:
-        return BLE_ERROR_INVALID_STATE;
-    case NRF_ERROR_BUSY:
-        return BLE_ERROR_BUSY;
-    case NRF_ERROR_NO_MEM:
-        return BLE_ERROR_NO_MEMORY;
-    default:
-        return BLE_ERROR_UNKNOWN;
-    }
-}
-
-int updateConnInterval(uint16_t connHandle, uint16_t interval) {
-    ble_gap_conn_params_t param = {};
-    param.min_conn_interval = interval;
-    param.max_conn_interval = interval;
-    param.slave_latency = SLAVE_LATENCY;
-    param.conn_sup_timeout = CONN_SUP_TIMEOUT;
-    LOG_DEBUG(TRACE, "Requesting connection interval: %u x 1.25 ms", (unsigned)interval);
-    const auto ret = sd_ble_gap_conn_param_update(connHandle, &param);
-    if (ret != NRF_SUCCESS) {
-        LOG_DEBUG(WARN, "sd_ble_gap_conn_param_update() failed: %u", (unsigned)ret);
-        return halError(ret);
+    uint8_t discoverAll_ : 1;                           /**< If it is going to discover all service during the service discovery procedure. */
+    uint8_t isDiscovering_ : 1;                         /**< If there is on-going discovery procedure. */
+    uint16_t currDiscConnHandle_;                       /**< Current connection handle under which the service and characteristics to be discovered. */
+    uint8_t currDiscProcedure_;                         /**< Current discovery procedure. */
+    hal_ble_svc_t currDiscSvc_;                         /**< Current service to be discovered for the characteristics. */
+    os_semaphore_t discoverySemaphore_;                 /**< Semaphore to wait until the discovery procedure completed. */
+    os_semaphore_t readWriteSemaphore_;                 /**< Semaphore to wait until the read or write operation completed. */
+    Vector<hal_ble_svc_t> discoveredServices_;          /**< Discover services. */
+    Vector<hal_ble_char_t> discoveredCharacteristics_;  /**< Discovered characteristics. */
+    uint16_t readAttrHandle_;                           /**< Current handle of which attribute to be read. */
+    uint8_t* readBuf_;                                  /**< Current buffer to be filled for the read data. */
+    uint16_t readLen_;                                  /**< Length of read data. */
+    BleObject* bleObj_;                                 /**< Pointer to the BLE Object. */
+    AtomicAllocedPool gattcDiscSvcCharPool_;            /**< Pool to allocate memory for discovered services or characteristics. */
+    AtomicAllocedPool gattcDataRecPool_;                /**< Pool to allocate memory for received data. */
+    uint16_t attMtuExchangeConnHandle_;                 /**< Current handle of the connection to execute ATT_MTU exchange procedure. */
+    os_timer_t attMtuExchangeTimer_;                    /**< Timer used for sending the exchanging ATT_MTU request after connection established. */
+
+    static const uint16_t SERVICES_BASE_START_HANDLE = 0x0001;
```

Comment:
Consider moving these out of class into the anonymous namespace in the top of the source file for better visibility.